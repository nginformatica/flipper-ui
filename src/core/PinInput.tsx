import React, { useEffect, useRef } from 'react'
import { TextField } from '@material-ui/core'

interface PinInputGridProps {
    pin: Array<number | undefined>
    setPin: (pin: Array<number | undefined>) => void
    onPinChanged: (pinEntry: number | undefined, index: number) => void
    pinLength: number
    validationResult: boolean | undefined
    isValidating: boolean
    size: 'small' | 'large'
}

const removeValuesFromArray = (valuesArray: string[], value: string) => {
    const valueIndex = valuesArray.findIndex(entry => entry === value)
    if (valueIndex === -1) {
        return
    }
    valuesArray.splice(valueIndex, 1)
}

const PIN_MIN_VALUE = 0
const PIN_MAX_VALUE = 9
const BACKSPACE_KEY = 'Backspace'

const PinInput: React.FC<PinInputGridProps> = ({
    pinLength,
    pin,
    setPin,
    onPinChanged,
    validationResult,
    isValidating,
    size
}) => {
    const inputRefs = useRef<HTMLInputElement[]>([])

    const changePinFocus = (pinIndex: number) => {
        const ref = inputRefs.current[pinIndex]
        if (ref) {
            ref.focus()
        }
    }

    useEffect(() => {
        changePinFocus(0)
    }, [isValidating])

    const onPaste = (event: React.ClipboardEvent<HTMLInputElement>) => {
        const pastedValue = event.clipboardData.getData('text/plain')
        const splitValues = pastedValue.split('')
        const intValues: number[] = []
        for (const value of splitValues) {
            intValues.push(parseInt(value))
        }
        setPin(intValues)
        changePinFocus(intValues.length - 1)
    }

    const onChange = (
        event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
        index: number
    ) => {
        if (event.target.value.length > 1) {return}
        const previousValue = event.target.defaultValue
        const valuesArray = event.target.value.split('')
        removeValuesFromArray(valuesArray, previousValue)
        const value = valuesArray.pop()
        if (!value) {
            return
        }
        const pinNumber = Number(value.trim())
        if (isNaN(pinNumber) || value.length === 0) {
            return
        }

        if (pinNumber >= PIN_MIN_VALUE && pinNumber <= PIN_MAX_VALUE) {
            onPinChanged(pinNumber, index)
            if (index < pinLength - 1) {
                changePinFocus(index + 1)
            }
        }
    }

    const onKeyDown = (
        event: React.KeyboardEvent<HTMLDivElement>,
        index: number
    ) => {
        const keyboardKeyCode = event.nativeEvent.code
        if (keyboardKeyCode !== BACKSPACE_KEY) {
            return
        }

        if (pin[index] === undefined) {
            changePinFocus(index - 1)
        } else {
            onPinChanged(undefined, index)
        }
    }

    return (
        <>
            <div>
                { Array.from({ length: pinLength }, (_, index) => (
                    <TextField
                        disabled={ isValidating }
                        variant='outlined'
                        onKeyDown={ event => onKeyDown(event, index) }
                        onPaste={ onPaste }
                        color='primary'
                        error={ validationResult }
                        style={ {
                            width: size === 'small' ? '40px' : '40px',
                            height: size === 'small' ? '30px' : '40px',
                            marginInline: size === 'small' ? '5px' : '10px'
                        } }
                        InputProps={ {
                            style: {
                                width: size === 'small' ? '40px' : '45px',
                                textAlign: 'center',
                                fontWeight: 'bold',
                                fontSize: size === 'small' ? '16px' : '20px',
                                padding: 'auto'
                            }
                        } }
                        inputRef={ (el: HTMLInputElement) => {
                            inputRefs.current[index] = el
                        } }
                        key={ index }
                        onChange={ event => {
                            onChange(event, index)
                        } }
                        value={ pin[index] || '' }
                    />
                )) }
            </div>
        </>
    )
}

export default PinInput
