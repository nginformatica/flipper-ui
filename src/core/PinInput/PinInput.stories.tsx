import React, { useState } from 'react'
import { Meta } from '@storybook/react'
import PinInput from '.'
import { Button } from '@material-ui/core'
import styled from 'styled-components'

export default {
    title: 'PinInput',
    component: PinInput
} as Meta<typeof PinInput>

const ValidateContainer = styled.div`
    display: flex;
    width: auto;
    justify-content: center;
    margin-bottom: 2rem;
`

const ButtonContainer = styled.div`
    display: flex;
    width: auto;
    justify-content: center;
    margin-top: 3rem;
`

const PIN_LENGTH = 6

export const Default = () => {
    const [pin, setPin] = useState<Array<number | undefined>>(
        new Array(PIN_LENGTH)
    )

    const onPinChanged = (pinEntry: number | undefined, index: number) => {
        const newPin = [...pin]
        newPin[index] = pinEntry
        setPin(newPin)
    }
    const [hasError, setHasError] = useState(false)
    const [isValidating, setIsValidating] = useState(false)

    const handleValidation = () => {
        setIsValidating(true)
        if (pin.join('') !== '123123') {
            setHasError(true)
            setPin(new Array(PIN_LENGTH))
        } else {
            setHasError(false)
            alert('PIN is correct')
        }
        setIsValidating(false)
    }

    return (
        <>
            <ValidateContainer>
                <span>Valid pin: 123123</span>
            </ValidateContainer>
            <PinInput
                pin={pin}
                setPin={setPin}
                onPinChanged={onPinChanged}
                pinLength={PIN_LENGTH}
                validationResult={hasError}
                isValidating={isValidating}
                size='small'
                variant='outlined'
            />
            <ButtonContainer>
                <Button
                    onClick={() => {
                        handleValidation()
                    }}>
                    Validate
                </Button>
            </ButtonContainer>
        </>
    )
}

export const Large = () => {
    const [pin, setPin] = useState<Array<number | undefined>>(
        new Array(PIN_LENGTH)
    )

    const onPinChanged = (pinEntry: number | undefined, index: number) => {
        const newPin = [...pin]
        newPin[index] = pinEntry
        setPin(newPin)
    }
    const [hasError, setHasError] = useState(false)
    const [isValidating, setIsValidating] = useState(false)

    const handleValidation = () => {
        setIsValidating(true)
        if (pin.join('') !== '123123') {
            setPin(new Array(PIN_LENGTH))
            setHasError(true)
        } else {
            setHasError(false)
            alert('PIN is correct')
        }
        setIsValidating(false)
    }

    return (
        <>
            <ValidateContainer>
                <span>Valid pin: 123123</span>
            </ValidateContainer>
            <PinInput
                pin={pin}
                setPin={setPin}
                onPinChanged={onPinChanged}
                pinLength={PIN_LENGTH}
                validationResult={hasError}
                isValidating={isValidating}
                size='large'
            />
            <ButtonContainer>
                <Button
                    onClick={() => {
                        handleValidation()
                    }}>
                    Validate
                </Button>
            </ButtonContainer>
        </>
    )
}
