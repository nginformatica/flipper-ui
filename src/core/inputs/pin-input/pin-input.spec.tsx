import React from 'react'
import { act } from 'react-dom/test-utils'
import { fireEvent, render, screen } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'
import MockedComponent from '@/test/mocks/pin-input-mock'
import PinInput from '.'

describe('PinInput', () => {
    it('should render', () => {
        render(
            <PinInput
                isValidating={false}
                pin={[]}
                pinLength={6}
                setPin={jest.fn()}
                size='small'
                validationResult={false}
                onPinChanged={jest.fn()}
            />
        )

        const pinInput = screen.getAllByRole('pin-input-field')

        expect(pinInput).toHaveLength(6)
    })

    it('should update zeros', async () => {
        render(<MockedComponent />)

        const pinInput = screen.getAllByRole('pin-input-field')

        fireEvent.focus(pinInput[0])

        await act(async () => {
            await userEvent.type(pinInput[0], '012345')
        })

        const pinInputAfterType = screen.getAllByRole('pin-input-field')

        const pin = pinInputAfterType
            .map(inputContainer => {
                return (inputContainer.firstChild as HTMLInputElement).value
            })
            .join('')

        expect(pin).toEqual('012345')
    })

    it('should update on typing', async () => {
        render(<MockedComponent />)

        const pinInput = screen.getAllByRole('pin-input-field')

        fireEvent.focus(pinInput[0])

        await act(async () => {
            await userEvent.type(pinInput[0], '123456')
        })

        const pinInputAfterType = screen.getAllByRole('pin-input-field')

        const pin = pinInputAfterType
            .map(inputContainer => {
                return (inputContainer.firstChild as HTMLInputElement).value
            })
            .join('')

        expect(pin).toEqual('123456')
    })

    it('should do nothing when typing when input is full', async () => {
        render(<MockedComponent />)

        const pinInput = screen.getAllByRole('pin-input-field')

        fireEvent.focus(pinInput[0])

        await act(async () => {
            await userEvent.type(pinInput[0], '1')
            await userEvent.type(pinInput[1], '2')
            await userEvent.type(pinInput[2], '3')
            await userEvent.type(pinInput[3], '4')
            await userEvent.type(pinInput[4], '5')
            await userEvent.type(pinInput[5], '6')
            await userEvent.type(pinInput[5], '7')
        })

        const pinInputAfterType = screen.getAllByRole('pin-input-field')

        const pin = pinInputAfterType
            .map(inputContainer => {
                return (inputContainer.firstChild as HTMLInputElement).value
            })
            .join('')

        expect(pin).toEqual('123456')
    })

    it('should update on typing with backspace', async () => {
        render(<MockedComponent />)

        const pinInput = screen.getAllByRole('pin-input-field')

        fireEvent.focus(pinInput[0])

        await act(async () => {
            await userEvent.type(pinInput[0], '1234{backspace}{backspace}356')
        })

        const pinInputAfterType = screen.getAllByRole('pin-input-field')

        const pin = pinInputAfterType
            .map(inputContainer => {
                return (inputContainer.firstChild as HTMLInputElement).value
            })
            .join('')

        expect(pin).toEqual('123356')
    })

    it('should update correct if pasted a pin with larger size', async () => {
        render(<MockedComponent />)

        const pinInput = screen.getAllByRole('pin-input-field')

        await act(async () => {
            fireEvent.paste(pinInput[0], {
                clipboardData: {
                    getData: () => '123456789'
                }
            })
        })

        const pinInputAfterType = screen.getAllByRole('pin-input-field')

        const pin = pinInputAfterType
            .map(inputContainer => {
                return (inputContainer.firstChild as HTMLInputElement).value
            })
            .join('')

        expect(pin).toEqual('123456')
    })

    it('should not update when typing strings', async () => {
        render(<MockedComponent />)

        const pinInput = screen.getAllByRole('pin-input-field')

        fireEvent.focus(pinInput[0])

        await act(async () => {
            await userEvent.type(pinInput[0], '123y56')
        })

        const pinInputAfterType = screen.getAllByRole('pin-input-field')

        const pin = pinInputAfterType
            .map(inputContainer => {
                return (inputContainer.firstChild as HTMLInputElement).value
            })
            .join('')

        expect(pin).toEqual('12356')
    })

    it('should update on typing and removing with backspace', async () => {
        render(<MockedComponent />)

        const pinInput = screen.getAllByRole('pin-input-field')

        fireEvent.focus(pinInput[0])

        await act(async () => {
            await userEvent.type(pinInput[0], '123456')
            await userEvent.type(pinInput[pinInput.length - 1], '{backspace}')
        })

        const pinInputAfterType = screen.getAllByRole('pin-input-field')

        const pin = pinInputAfterType
            .map(inputContainer => {
                return (inputContainer.firstChild as HTMLInputElement).value
            })
            .join('')

        expect(pin).toEqual('12345')
    })

    it('should do nothing on typing backspace with empty pin', async () => {
        const { container } = render(<MockedComponent />)

        const pinInput = screen.getAllByRole('pin-input-field')

        const prev = container.innerHTML

        await act(async () => {
            await userEvent.type(pinInput[0], '{backspace}')
        })

        const after = container.innerHTML

        expect(prev).toEqual(after)
    })

    it('should update on paste', async () => {
        render(<MockedComponent />)

        const pinInput = screen.getAllByRole('pin-input-field')

        await act(async () => {
            fireEvent.paste(pinInput[0], {
                clipboardData: {
                    getData: () => '123456'
                }
            })
        })

        const pinInputAfterType = screen.getAllByRole('pin-input-field')

        const pin = pinInputAfterType
            .map(inputContainer => {
                return (inputContainer.firstChild as HTMLInputElement).value
            })
            .join('')

        expect(pin).toEqual('123456')
    })

    it('should trim pin on paste', async () => {
        render(<MockedComponent />)

        const pinInput = screen.getAllByRole('pin-input-field')

        await act(async () => {
            fireEvent.paste(pinInput[0], {
                clipboardData: {
                    getData: () => ' 123456 '
                }
            })
        })

        const pinInputAfterType = screen.getAllByRole('pin-input-field')

        const pin = pinInputAfterType
            .map(inputContainer => {
                return (inputContainer.firstChild as HTMLInputElement).value
            })
            .join('')

        expect(pin).toEqual('123456')
    })

    it('should render on large size', () => {
        render(<MockedComponent pinInputProps={{ size: 'large' }} />)

        const pinInput = screen.getAllByRole('pin-input-field')[0]
        const container = pinInput.parentElement

        expect(pinInput).toHaveProperty('style.width', '45px')
        expect(pinInput).toHaveProperty('style.font-size', '20px')
        expect(container).toHaveProperty('style.width', '40px')
        expect(container).toHaveProperty('style.height', '40px')
        expect(container).toHaveProperty('style.margin-inline', '10px')
    })

    it('should match snapshot', () => {
        const { container } = render(
            <PinInput
                isValidating={false}
                pin={[]}
                pinLength={6}
                setPin={jest.fn()}
                size='small'
                validationResult={false}
                onPinChanged={jest.fn()}
            />
        )

        expect(container).toMatchSnapshot()
    })
})
