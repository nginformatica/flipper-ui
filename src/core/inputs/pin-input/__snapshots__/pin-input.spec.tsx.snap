// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PinInput should match snapshot 1`] = `
<div>
  <div
    class="sc-aYaIB lNsDf"
  >
    <div
      class="MuiFormControl-root MuiTextField-root pin-input-field"
      style="width: 40px; height: 30px; margin-inline: 5px;"
    >
      <div
        class="MuiInputBase-root MuiOutlinedInput-root Mui-focused Mui-focused MuiInputBase-formControl"
        role="pin-input-field"
        style="width: 40px; text-align: center; font-weight: bold; font-size: 16px;"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-49 MuiOutlinedInput-notchedOutline"
          style="padding-left: 8px;"
        >
          <legend
            class="PrivateNotchedOutline-legend-50"
            style="width: 0.01px;"
          >
            <span>
              ​
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root pin-input-field"
      style="width: 40px; height: 30px; margin-inline: 5px;"
    >
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        role="pin-input-field"
        style="width: 40px; text-align: center; font-weight: bold; font-size: 16px;"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-49 MuiOutlinedInput-notchedOutline"
          style="padding-left: 8px;"
        >
          <legend
            class="PrivateNotchedOutline-legend-50"
            style="width: 0.01px;"
          >
            <span>
              ​
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root pin-input-field"
      style="width: 40px; height: 30px; margin-inline: 5px;"
    >
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        role="pin-input-field"
        style="width: 40px; text-align: center; font-weight: bold; font-size: 16px;"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-49 MuiOutlinedInput-notchedOutline"
          style="padding-left: 8px;"
        >
          <legend
            class="PrivateNotchedOutline-legend-50"
            style="width: 0.01px;"
          >
            <span>
              ​
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root pin-input-field"
      style="width: 40px; height: 30px; margin-inline: 5px;"
    >
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        role="pin-input-field"
        style="width: 40px; text-align: center; font-weight: bold; font-size: 16px;"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-49 MuiOutlinedInput-notchedOutline"
          style="padding-left: 8px;"
        >
          <legend
            class="PrivateNotchedOutline-legend-50"
            style="width: 0.01px;"
          >
            <span>
              ​
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root pin-input-field"
      style="width: 40px; height: 30px; margin-inline: 5px;"
    >
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        role="pin-input-field"
        style="width: 40px; text-align: center; font-weight: bold; font-size: 16px;"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-49 MuiOutlinedInput-notchedOutline"
          style="padding-left: 8px;"
        >
          <legend
            class="PrivateNotchedOutline-legend-50"
            style="width: 0.01px;"
          >
            <span>
              ​
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root pin-input-field"
      style="width: 40px; height: 30px; margin-inline: 5px;"
    >
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        role="pin-input-field"
        style="width: 40px; text-align: center; font-weight: bold; font-size: 16px;"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-49 MuiOutlinedInput-notchedOutline"
          style="padding-left: 8px;"
        >
          <legend
            class="PrivateNotchedOutline-legend-50"
            style="width: 0.01px;"
          >
            <span>
              ​
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
